File: game/tv_show/playing.py
from forest.forest_game import ForestGame
from game_data import GameData
from phone_events import PhoneEvents
from state import State
from tv_show.attract import Attract
from tv_show.in_scoreboard import InScoreboard


class Playing(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.forest = ForestGame(context)

    def process_events(self, phone_events: PhoneEvents):
        if phone_events.hungup:
            return Attract

        if self.forest.ended:
            return InScoreboard

        return self.forest.process_events(phone_events)

    def render(self, screen):
        self.forest.render(screen)

    def on_exit(self):
        super().on_exit()
        if self.forest is not None and not self.forest.ended:
            self.forest.end()
        self.forest = None

    def external_effect(self, effect):
        if self.forest is not None:
            self.forest.external_effect(effect)

File: game-web/src/tvShow/playing.ts
/** Port of game/tv_show/playing.py */

import { State, type StateClass } from '../state';
import { ForestGame } from '../forest/forestGame';
import { Attract } from './attract';
import { InScoreboard } from './inScoreboard';
import type { EffectType } from '../effectType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class Playing extends State {
  private forest: ForestGame | null;

  constructor(context: GameData) {
    super(context);
    this.forest = new ForestGame(context);
  }

  processEvents(events: PhoneEvents): StateClass | null {
    if (events.hungup) return Attract;
    if (this.forest!.ended) return InScoreboard;
    this.forest!.processEvents(events);
    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    this.forest!.render(ctx);
  }

  onExit(): void {
    super.onExit();
    if (this.forest !== null && !this.forest.ended) {
      this.forest.end();
    }
    this.forest = null;
  }

  externalEffect(effect: EffectType): void {
    if (this.forest !== null) {
      this.forest.externalEffect(effect);
    }
  }
}
